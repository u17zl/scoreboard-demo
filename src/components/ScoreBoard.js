import React, { Fragment, useState, useEffect } from "react";
import { useSpring, animated } from "react-spring";
import { MainSection, TeamStat } from "./components";

const ScoreBoard = props => {
  const { home, away, score, stat, animation } = props;
  const [mainShow, setMainShow] = useState(false);
  const [statShow, setStatShow] = useState(false);

  //   input error handler
  useEffect(() => {
    switch (animation.length) {
      case 0:
        setMainShow(false);
        setStatShow(false);
        break;

      case 1:
        if (animation[0]["animation"] === "teamStat") {
          break;
        }
        console.log("hello");
        setMainShow(true);
        setStatShow(false);
        break;

      case 2:
        if (
          animation[0]["animation"] === "teamStat" &&
          animation[1]["animation"] === "main"
        ) {
          setMainShow(true);
          setStatShow(true);
          break;
        } else if (
          animation[1]["animation"] === "teamStat" &&
          animation[0]["animation"] === "main"
        ) {
          setMainShow(true);
          setStatShow(true);
          break;
        }

      default:
        break;
    }
  }, [props]);

  return (
    <Fragment>
      {console.log(mainShow)}
      {mainShow ? (
        <MainSection home={home} away={away} score={score} />
      ) : (
        <div style={{ height: 21 }} />
      )}
      {statShow ? <TeamStat stat={stat} /> : null}
    </Fragment>
  );
};

export default ScoreBoard;
