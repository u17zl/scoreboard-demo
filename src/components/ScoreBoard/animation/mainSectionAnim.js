import posed from "react-pose";

export const Home = posed.div({
  enter: {
    width: "100%",
    opacity: 1,
    delay: ({ delay }) => delay + 100,
    transition: { duration: 700 }
  },
  exit: {
    opacity: 0,
    width: 30,
    delay: 500,
    transition: { duration: 600 }
  }
});

export const Score = posed.div({
  enter: {
    width: "100%",
    opacity: 1,
    delay: ({ delay }) => delay + 0,
    transition: { duration: 300 }
  },
  exit: {
    width: 30,
    opacity: 0,
    delay: 500,
    ease: "e",
    transition: { duration: 600 }
  }
});
export const Away = posed.div({
  enter: {
    width: "100%",
    opacity: 1,
    delay: ({ delay }) => delay + 100,
    transition: { duration: 700 }
  },
  exit: {
    opacity: 0,
    width: 30,
    delay: 500,
    transition: { duration: 600 }
  }
});
export const Time = posed.div({
  enter: {
    width: "50%",
    opacity: 1,
    delay: ({ delay }) => delay + 900,
    transition: { duration: 500 }
  },
  exit: {
    opacity: 0,
    width: 30,
    transition: { duration: 600 }
  }
});

export const TimeText = posed.span({
  enter: {
    opacity: 1,
    delay: ({ delay }) => delay + 1050,
    transition: { duration: 350 }
  },
  exit: {
    opacity: 0,
    width: 30,
    transition: { duration: 600 }
  }
});

export const ScoreText = posed.div({
  enter: {
    opacity: 1,
    delay: ({ delay }) => delay + 100,
    transition: { duration: 500 },
    fontSize: 24
  },
  exit: {
    opacity: 0,
    transition: { duration: 600 }
  }
});

export const HomeText = posed.span({
  enter: {
    opacity: 1,
    delay: ({ delay }) => delay + 300,
    transition: { duration: 200 },
    fontSize: 24
  },
  exit: {
    opacity: 0,
    transition: { duration: 600 }
  }
});

export const AwayText = posed.div({
  enter: {
    opacity: 1,
    delay: ({ delay }) => delay + 300,
    transition: { duration: 200 },
    fontSize: 24
  },
  exit: {
    opacity: 0,
    transition: { duration: 600 }
  }
});
export const HomeBorder = posed.span({
  enter: {
    opacity: 0.9,
    delay: ({ delay }) => delay + 200,
    width: 7,
    backgroundColor: ({ color }) => color,
    transition: { duration: 300 }
  },
  exit: {
    opacity: 0,
    transition: { duration: 600 }
  }
});

export const AwayBorder = posed.span({
  enter: {
    opacity: 0.9,
    delay: ({ delay }) => delay + 200,
    width: 7,
    backgroundColor: ({ color }) => color,
    transition: { duration: 300 }
  },
  exit: {
    opacity: 0,
    transition: { duration: 600 }
  }
});
