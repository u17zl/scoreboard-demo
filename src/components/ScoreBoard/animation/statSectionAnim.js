import posed from "react-pose";

// stat total animation time team(400-1000) + stat(1000-1500)
export const Team = posed.div({
  enter: {
    width: "100%",
    opacity: 1,
    delay: ({ delay }) => delay + 500,
    transition: { duration: 500 }
  },
  exit: {
    opacity: 0,
    width: 50,
    delay: ({ isBoth }) => (isBoth ? 0 : 600),
    transition: ({ isBoth }) => (isBoth ? { duration: 200 } : { duration: 400 })
  }
});
export const TeamText = posed.span({
  enter: {
    opacity: 1,
    delay: ({ delay }) => delay + 600,
    transition: { duration: 400 }
  },
  exit: {
    opacity: 0,
    delay: ({ isBoth }) => (isBoth ? 0 : 500),
    transition: ({ isBoth }) => (isBoth ? { duration: 200 } : { duration: 400 })
  }
});

export const Stat = posed.div({
  enter: {
    width: "100%",
    opacity: 1,
    delay: ({ delay }) => delay + 1000,
    transition: { duration: 500 }
  },
  exit: {
    opacity: 0,
    width: 30,
    delay: ({ isBoth }) => (isBoth ? 0 : 200),
    transition: { duration: 300 }
  }
});
export const StatText = posed.span({
  enter: {
    opacity: 1,
    delay: ({ delay }) => delay + 1200,
    transition: { duration: 300 }
  },
  exit: {
    opacity: 0,
    width: 30,
    transition: { duration: 300 }
  }
});
