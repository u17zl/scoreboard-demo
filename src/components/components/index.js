export { default as MainSection } from "./MainSection";
export { default as TeamStat } from "./TeamStat";
