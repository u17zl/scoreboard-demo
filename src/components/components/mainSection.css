.main-home {
  line-height: 2rem;
  font-size: 1.5rem;
  width: 10%;
  text-align: center;
  color: #fff;
  font-weight: 700;
}
.main-score {
  line-height: 2rem;
  font-size: 2rem;
  text-align: center;
  color: #fff;
  font-weight: 700;
}
.main-away {
  line-height: 2rem;
  font-size: 1.5rem;
  text-align: center;
  color: #000;
  width: 10%;
  font-weight: 700;
}
.main-time {
  line-height: 2rem;
  font-size: 1.5rem;
  width: 10%;
  text-align: center;
  color: #fff;
  font-weight: 700;
}
