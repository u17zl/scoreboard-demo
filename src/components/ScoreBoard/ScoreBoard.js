import React, { Fragment, useState, useEffect } from "react";
import { MainSection, TeamStat } from "./components";

const ScoreBoard = props => {
  const { home, away, score, stat, animation } = props;
  const [mainShow, setMainShow] = useState(false);
  const [statShow, setStatShow] = useState(false);
  const [isBoth, setIsBoth] = useState(false);
  const [mainDelay, setMainDelay] = useState(0);
  const [statDelay, setStatDelay] = useState(0);

  //   input error handler
  useEffect(() => {
    switch (animation.length) {
      case 0:
        setIsBoth(true);
        setStatShow(false);
        setMainShow(false);
        break;

      case 1:
        if (animation[0]["animation"] === "teamStat") {
          break;
        }
        setIsBoth(false);
        setMainShow(true);
        setStatShow(false);

        setMainDelay(animation[0]["delay"]);
        break;

      case 2:
        if (
          animation[0]["animation"] === "teamStat" &&
          animation[1]["animation"] === "main"
        ) {
          setIsBoth(false);
          setMainShow(true);
          setStatShow(true);
          setMainDelay(animation[1]["delay"]);
          setStatDelay(animation[0]["delay"]);
          break;
        } else if (
          animation[1]["animation"] === "teamStat" &&
          animation[0]["animation"] === "main"
        ) {
          setIsBoth(false);
          setMainShow(true);
          setStatShow(true);
          setMainDelay(animation[0]["delay"]);
          setStatDelay(animation[1]["delay"]);
          break;
        }

      default:
        break;
    }
  }, [props, isBoth]);

  return (
    <Fragment>
      <div className="board-grid">
        <MainSection
          show={mainShow}
          statShow={statShow}
          home={home}
          away={away}
          score={score}
          statShow={statShow}
          delay={mainDelay * 1000}
        />
        <TeamStat
          show={statShow}
          stat={stat}
          home={home}
          away={away}
          delay={statDelay * 1000 + mainDelay * 1000}
          isBoth={isBoth}
        />
      </div>
    </Fragment>
  );
};

export default ScoreBoard;
