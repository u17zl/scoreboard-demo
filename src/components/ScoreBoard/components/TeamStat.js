import React, { Fragment, useState, useEffect } from "react";
import "../styles/teamStat.scss";
import posed, { PoseGroup } from "react-pose";
import * as statAnim from "../animation/statSectionAnim";

const { Team, TeamText, Stat, StatText } = statAnim;
const TeamStat = props => {
  const { show, home, away, stat, delay, isBoth } = props;
  const [statData, setSataData] = useState({ name: "", stat: "", color: "" });
  const { color: homeColor, name: homeName, abbr: homeAbbr } = home;
  const { color: awayColor, name: awayName, abbr: awayAbbr } = away;

  useEffect(() => {
    if (Object.keys(stat)[0] === "home") {
      setSataData({ name: homeName, stat: stat["home"], color: homeColor });
    } else {
      setSataData({ name: awayName, stat: stat["away"], color: awayColor });
    }
  }, [props]);

  return (
    <Fragment>
      <PoseGroup>
        {show && [
          <Team
            key="stat"
            className="team-stat"
            style={{ backgroundColor: statData.color }}
            delay={delay}
            isBoth={isBoth}
          >
            <TeamText delay={delay} isBoth={isBoth}>
              {statData.name}
            </TeamText>
          </Team>,
          <Stat
            key="stat-info"
            className="info-stat"
            style={{ backgroundColor: "#001339" }}
            delay={delay}
            isBoth={isBoth}
          >
            <StatText key="stat-text" delay={delay} isBoth={isBoth}>
              {statData.stat}
            </StatText>
          </Stat>
        ]}
      </PoseGroup>
    </Fragment>
  );
};

export default TeamStat;
