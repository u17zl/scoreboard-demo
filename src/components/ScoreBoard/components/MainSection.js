import React, { Fragment } from "react";
import { PoseGroup } from "react-pose";
import "../styles/mainSection.scss";
import * as mainAnim from "../animation/mainSectionAnim";

const {
  Home,
  HomeText,
  HomeBorder,
  Score,
  ScoreText,
  Away,
  AwayBorder,
  AwayText,
  Time,
  TimeText
} = mainAnim;

const MainSection = props => {
  const { show, statShow, home, away, score, delay } = props;

  const {
    color: homeColor,
    name: homeName,
    abbr: homeAbbr,
    borderColor: homeBorderColor
  } = home;
  const {
    color: awayColor,
    name: awayName,
    abbr: awayAbbr,
    borderColor: awayBorderColor
  } = away;
  const { home: homeScore, away: awayScore } = score;

  return (
    <Fragment>
      <PoseGroup>
        {show && [
          <Home
            key="home"
            className="main-home"
            style={{ backgroundColor: homeColor }}
            delay={delay}
          >
            <HomeText key="texthome" delay={delay}>
              {homeAbbr}
            </HomeText>
            <HomeBorder
              key="borderhome"
              className="border-home"
              color={homeBorderColor}
              delay={delay}
            />
          </Home>,
          <Score
            key="score"
            className="main-score"
            style={{ backgroundColor: "#001339" }}
            delay={delay}
          >
            <ScoreText
              key="textscore"
              delay={delay}
            >{`${homeScore}-${awayScore}`}</ScoreText>
          </Score>,
          <Away
            key="away"
            className="main-away"
            style={{ backgroundColor: awayColor }}
            delay={delay}
          >
            <AwayBorder
              key="borderaway"
              className="border-away"
              color={awayBorderColor}
              delay={delay}
            />
            <AwayText key="textaway" delay={delay}>
              {awayAbbr}
            </AwayText>
          </Away>
        ]}

        {show && !statShow && (
          <Time
            key="time"
            className="main-time"
            style={{ backgroundColor: "#001339" }}
            delay={delay}
          >
            <TimeText key="time-text">90:00</TimeText>
          </Time>
        )}
      </PoseGroup>
    </Fragment>
  );
};

export default MainSection;
