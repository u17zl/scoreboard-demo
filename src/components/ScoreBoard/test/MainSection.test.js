import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import ScoreBoard from "../ScoreBoard";
import { MainSection } from "../components";
import Adapter from "enzyme-adapter-react-16";
import * as mainAnim from "../animation/mainSectionAnim";

const {
  Home,
  HomeText,
  HomeBorder,
  Score,
  ScoreText,
  Away,
  AwayBorder,
  AwayText,
  Time,
  TimeText
} = mainAnim;

Enzyme.configure({ adapter: new Adapter() });

const homeInfoData = {
  name: "Kingston City",
  abbr: "KC",
  color: "#d6211c",
  borderColor: "#de4d49"
};

const awayInfoData = {
  name: "Green Gully",
  abbr: "GG",
  color: "#eee",
  borderColor: "#ccc"
};

const scoreData = { home: 4, away: 0 };

const statData = { home: "Red Cards 0" };

const setupMainShow = () => {
  const props = {
    show: true,
    statShow: false,
    home: homeInfoData,
    away: awayInfoData,
    stat: statData,
    score: scoreData,
    delay: 0
  };
  const sWrapper = shallow(<MainSection {...props} />);
  const mWrapper = mount(<MainSection {...props} />);
  return {
    props,
    sWrapper,
    mWrapper
  };
};

const setupMainHide = () => {
  const props = {
    show: false,
    statShow: false,
    home: homeInfoData,
    away: awayInfoData,
    stat: statData,
    score: scoreData,
    delay: 0
  };
  const sWrapper = shallow(<MainSection {...props} />);
  const mWrapper = mount(<MainSection {...props} />);
  return {
    props,
    sWrapper,
    mWrapper
  };
};

const setupMainShowToHide = () => {
  const props = {
    show: true,
    statShow: true,
    home: homeInfoData,
    away: awayInfoData,
    stat: statData,
    score: scoreData,
    delay: 0
  };
  const sWrapper = shallow(<MainSection {...props} />);
  const mWrapper = mount(<MainSection {...props} />);
  return {
    props,
    sWrapper,
    mWrapper
  };
};

describe("Main Sector show", () => {
  const { sWrapper, mWrapper, props } = setupMainShow();

  it("get child component MainSectionShow's length", () => {
    expect(sWrapper.find(Home).length).toBe(1);
    expect(sWrapper.find(HomeText).length).toBe(1);
    expect(sWrapper.find(HomeBorder).length).toBe(1);
    expect(sWrapper.find(Score).length).toBe(1);
    expect(sWrapper.find(ScoreText).length).toBe(1);
    expect(sWrapper.find(Away).length).toBe(1);
    expect(sWrapper.find(AwayBorder).length).toBe(1);
    expect(sWrapper.find(AwayText).length).toBe(1);
    expect(sWrapper.find(Time).length).toBe(1);
    expect(sWrapper.find(TimeText).length).toBe(1);

    expect(mWrapper.find(Home).length).toBe(1);
    expect(mWrapper.find(HomeText).length).toBe(1);
    expect(mWrapper.find(HomeBorder).length).toBe(1);
    expect(mWrapper.find(Score).length).toBe(1);
    expect(mWrapper.find(ScoreText).length).toBe(1);
    expect(mWrapper.find(Away).length).toBe(1);
    expect(mWrapper.find(AwayBorder).length).toBe(1);
    expect(mWrapper.find(AwayText).length).toBe(1);
    expect(mWrapper.find(Time).length).toBe(1);
    expect(mWrapper.find(TimeText).length).toBe(1);
  });

  // Snapshot
  it("Main Section Show Snapshot", () => {
    expect(toJson(sWrapper)).toMatchSnapshot();
  });
});

describe("Main Sector hide", () => {
  const { sWrapper, mWrapper, props } = setupMainHide();

  it("get child component MainSectionHide's length", () => {
    expect(sWrapper.find(Home).length).toBe(0);
    expect(sWrapper.find(HomeText).length).toBe(0);
    expect(sWrapper.find(HomeBorder).length).toBe(0);
    expect(sWrapper.find(Score).length).toBe(0);
    expect(sWrapper.find(ScoreText).length).toBe(0);
    expect(sWrapper.find(Away).length).toBe(0);
    expect(sWrapper.find(AwayBorder).length).toBe(0);
    expect(sWrapper.find(AwayText).length).toBe(0);
    expect(sWrapper.find(Time).length).toBe(0);
    expect(sWrapper.find(TimeText).length).toBe(0);

    expect(mWrapper.find(Home).length).toBe(0);
    expect(mWrapper.find(HomeText).length).toBe(0);
    expect(mWrapper.find(HomeBorder).length).toBe(0);
    expect(mWrapper.find(Score).length).toBe(0);
    expect(mWrapper.find(ScoreText).length).toBe(0);
    expect(mWrapper.find(Away).length).toBe(0);
    expect(mWrapper.find(AwayBorder).length).toBe(0);
    expect(mWrapper.find(AwayText).length).toBe(0);
    expect(mWrapper.find(Time).length).toBe(0);
    expect(mWrapper.find(TimeText).length).toBe(0);
  });

  // Snapshot
  it("Main Section Hide Snapshot", () => {
    expect(toJson(sWrapper)).toMatchSnapshot();
  });
});

describe("Main Sector show then hide", () => {
  const { sWrapper, mWrapper, props } = setupMainShowToHide();

  it("get child component MainSectionShowToHide's length", () => {
    expect(sWrapper.find(Home).length).toBe(1);
    expect(sWrapper.find(HomeText).length).toBe(1);
    expect(sWrapper.find(HomeBorder).length).toBe(1);
    expect(sWrapper.find(Score).length).toBe(1);
    expect(sWrapper.find(ScoreText).length).toBe(1);
    expect(sWrapper.find(Away).length).toBe(1);
    expect(sWrapper.find(AwayBorder).length).toBe(1);
    expect(sWrapper.find(AwayText).length).toBe(1);
    expect(sWrapper.find(Time).length).toBe(0);
    expect(sWrapper.find(TimeText).length).toBe(0);

    expect(mWrapper.find(Home).length).toBe(1);
    expect(mWrapper.find(HomeText).length).toBe(1);
    expect(mWrapper.find(HomeBorder).length).toBe(1);
    expect(mWrapper.find(Score).length).toBe(1);
    expect(mWrapper.find(ScoreText).length).toBe(1);
    expect(mWrapper.find(Away).length).toBe(1);
    expect(mWrapper.find(AwayBorder).length).toBe(1);
    expect(mWrapper.find(AwayText).length).toBe(1);
    expect(mWrapper.find(Time).length).toBe(0);
    expect(mWrapper.find(TimeText).length).toBe(0);
  });

  // Snapshot
  it("Main Section Show To Hide Snapshot", () => {
    expect(toJson(sWrapper)).toMatchSnapshot();
  });
});
