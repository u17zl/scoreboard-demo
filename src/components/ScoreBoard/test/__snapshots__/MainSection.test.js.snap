// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Sector hide Main Section Hide Snapshot 1`] = `
<Fragment>
  <PoseGroup
    flipMove={true}
  />
</Fragment>
`;

exports[`Main Sector show Main Section Show Snapshot 1`] = `
<Fragment>
  <PoseGroup
    flipMove={true}
  >
    <ForwardRef
      className="main-home"
      delay={0}
      key="home"
      style={
        Object {
          "backgroundColor": "#d6211c",
        }
      }
    >
      <ForwardRef
        delay={0}
        key="texthome"
      >
        KC
      </ForwardRef>
      <ForwardRef
        className="border-home"
        color="#de4d49"
        delay={0}
        key="borderhome"
      />
    </ForwardRef>
    <ForwardRef
      className="main-score"
      delay={0}
      key="score"
      style={
        Object {
          "backgroundColor": "#001339",
        }
      }
    >
      <ForwardRef
        delay={0}
        key="textscore"
      >
        4-0
      </ForwardRef>
    </ForwardRef>
    <ForwardRef
      className="main-away"
      delay={0}
      key="away"
      style={
        Object {
          "backgroundColor": "#eee",
        }
      }
    >
      <ForwardRef
        className="border-away"
        color="#ccc"
        delay={0}
        key="borderaway"
      />
      <ForwardRef
        delay={0}
        key="textaway"
      >
        GG
      </ForwardRef>
    </ForwardRef>
    <ForwardRef
      className="main-time"
      delay={0}
      key="time"
      style={
        Object {
          "backgroundColor": "#001339",
        }
      }
    >
      <ForwardRef
        key="time-text"
      >
        90:00
      </ForwardRef>
    </ForwardRef>
  </PoseGroup>
</Fragment>
`;

exports[`Main Sector show then hide Main Section Show To Hide Snapshot 1`] = `
<Fragment>
  <PoseGroup
    flipMove={true}
  >
    <ForwardRef
      className="main-home"
      delay={0}
      key="home"
      style={
        Object {
          "backgroundColor": "#d6211c",
        }
      }
    >
      <ForwardRef
        delay={0}
        key="texthome"
      >
        KC
      </ForwardRef>
      <ForwardRef
        className="border-home"
        color="#de4d49"
        delay={0}
        key="borderhome"
      />
    </ForwardRef>
    <ForwardRef
      className="main-score"
      delay={0}
      key="score"
      style={
        Object {
          "backgroundColor": "#001339",
        }
      }
    >
      <ForwardRef
        delay={0}
        key="textscore"
      >
        4-0
      </ForwardRef>
    </ForwardRef>
    <ForwardRef
      className="main-away"
      delay={0}
      key="away"
      style={
        Object {
          "backgroundColor": "#eee",
        }
      }
    >
      <ForwardRef
        className="border-away"
        color="#ccc"
        delay={0}
        key="borderaway"
      />
      <ForwardRef
        delay={0}
        key="textaway"
      >
        GG
      </ForwardRef>
    </ForwardRef>
  </PoseGroup>
</Fragment>
`;
